src/Data/Json/JTable/Internal.purs

```diff
@@ -0,0 +1,218 @@
+module Data.Json.JTable.Internal
+  ( renderJTableRaw, renderRows, renderThead, renderTbody, tsToRows, sortTree
+  , Tree(..), tPath, tWidth, tHeight, tKids
+  , Cell(..), cCursor, cWidth, cHeight, cJsonPrim
+  , JPath(..), Table(..)
+  , tFromJson, tMergeArray, widthOfPrimTuple, padTree
+  , cFromJson, cMergeObj, mergeObjTuple
+  , foldJsonP, toPrim, zipWithIndex, orelse, strcmp, _nattr, _cspan, _rspan
+  ) where
+
+import Math (max)
+import Data.Either
+import Data.Maybe (Maybe(..), fromMaybe, isJust)
+import Data.Tuple
+import Data.String (joinWith, localeCompare)
+import Data.Array
+import Data.Foldable
+import qualified Data.StrMap as M
+import Data.Foldable (foldl, any, all, mconcat, elem)
+import Data.Traversable (for)
+import Control.Alt
+import Control.Apply
+import Control.Bind
+import Control.Monad
+import Control.MonadPlus
+
+import Data.Argonaut.Core
+import Data.Argonaut.JCursor
+import Data.Argonaut.Parser (jsonParser)
+import Text.Smolder.HTML (thead, tbody, tr, th, td)
+import Text.Smolder.Markup (Markup(..), Attribute(..), attribute, (!))
+
+-- path of object keys, with array indices omitted
+type JPath = [String]
+
+-- rows of cells
+type Table = [[Cell]]
+
+-- header data
+data Tree = T String JPath Number Number [Tree]
```

Comment:
It is true, the path `p` can be calculated top-down, and `w` and `h` can be calculated bottom-up. 
However, these are all produced when the tree is first built up, never updated.
The tree is traversed twice more, both top-down:
- to render the header in `toRows`
- to produce a table structure for the data in `cFromJson`
These two traversals don't need to pass the extra top-down and bottom-up state and calculate the results, because the data is already available. Besides not duplicating work this also saves stack space and code complexity. However, it does use somewhat more heap space, but I don't think that is a concern since the number of nodes is equal to the number of labels in the json.